From 3a0c1d6f3e20744b7daf25da285e31a506726b0b Mon Sep 17 00:00:00 2001
From: Lulu Cheng <chenglulu@loongson.cn>
Date: Wed, 12 Feb 2025 09:57:02 +0800
Subject: [PATCH 37/59] LoongArch: Move the function loongarch_register_pragmas
 to loongarch-c.cc.

gcc/ChangeLog:

	* config/loongarch/loongarch-target-attr.cc
	(loongarch_pragma_target_parse): Move to ...
	(loongarch_register_pragmas): Move to ...
	* config/loongarch/loongarch-c.cc
	(loongarch_pragma_target_parse): ... here.
	(loongarch_register_pragmas): ... here.
	* config/loongarch/loongarch-protos.h
	(loongarch_process_target_attr): Function Declaration.
---
 gcc/config/loongarch/loongarch-c.cc           | 51 +++++++++++++++++++
 gcc/config/loongarch/loongarch-protos.h       |  1 +
 gcc/config/loongarch/loongarch-target-attr.cc | 48 -----------------
 3 files changed, 52 insertions(+), 48 deletions(-)

diff --git a/gcc/config/loongarch/loongarch-c.cc b/gcc/config/loongarch/loongarch-c.cc
index 4ecea6a45..8c22196b3 100644
--- a/src/gcc/config/loongarch/loongarch-c.cc
+++ b/src/gcc/config/loongarch/loongarch-c.cc
@@ -23,9 +23,11 @@ along with GCC; see the file COPYING3.  If not see
 #include "config.h"
 #include "system.h"
 #include "coretypes.h"
+#include "target.h"
 #include "tm.h"
 #include "c-family/c-common.h"
 #include "cpplib.h"
+#include "tm_p.h"
 
 #define preprocessing_asm_p() (cpp_get_options (pfile)->lang == CLK_ASM)
 #define builtin_define(TXT) cpp_define (pfile, TXT)
@@ -134,3 +136,52 @@ loongarch_cpu_cpp_builtins (cpp_reader *pfile)
   builtin_define_with_int_value ("_LOONGARCH_SPFPSET", 32);
 
 }
+
+/* Hook to validate the current #pragma GCC target and set the state, and
+   update the macros based on what was changed.  If ARGS is NULL, then
+   POP_TARGET is used to reset the options.  */
+
+static bool
+loongarch_pragma_target_parse (tree args, tree pop_target)
+{
+  /* If args is not NULL then process it and setup the target-specific
+     information that it specifies.  */
+  if (args)
+    {
+      if (!loongarch_process_target_attr (args, NULL))
+	return false;
+
+      loongarch_option_override_internal (&la_target,
+					  &global_options,
+					  &global_options_set);
+    }
+
+  /* args is NULL, restore to the state described in pop_target.  */
+  else
+    {
+      pop_target = pop_target ? pop_target : target_option_default_node;
+      cl_target_option_restore (&global_options, &global_options_set,
+				TREE_TARGET_OPTION (pop_target));
+    }
+
+  target_option_current_node
+    = build_target_option_node (&global_options, &global_options_set);
+
+  loongarch_reset_previous_fndecl ();
+
+  /* If we're popping or reseting make sure to update the globals so that
+     the optab availability predicates get recomputed.  */
+  if (pop_target)
+    loongarch_save_restore_target_globals (pop_target);
+
+  return true;
+}
+
+/* Implement REGISTER_TARGET_PRAGMAS.  */
+
+void
+loongarch_register_pragmas (void)
+{
+  /* Update pragma hook to allow parsing #pragma GCC target.  */
+  targetm.target_option.pragma_parse = loongarch_pragma_target_parse;
+}
diff --git a/gcc/config/loongarch/loongarch-protos.h b/gcc/config/loongarch/loongarch-protos.h
index b76c9ce8c..b50b88585 100644
--- a/src/gcc/config/loongarch/loongarch-protos.h
+++ b/src/gcc/config/loongarch/loongarch-protos.h
@@ -219,4 +219,5 @@ extern void loongarch_option_override_internal (struct loongarch_target *, struc
 extern void loongarch_reset_previous_fndecl (void);
 extern void loongarch_save_restore_target_globals (tree new_tree);
 extern void loongarch_register_pragmas (void);
+extern bool loongarch_process_target_attr (tree args, tree fndecl);
 #endif /* ! GCC_LOONGARCH_PROTOS_H */
diff --git a/gcc/config/loongarch/loongarch-target-attr.cc b/gcc/config/loongarch/loongarch-target-attr.cc
index cee7031ca..cb537446d 100644
--- a/src/gcc/config/loongarch/loongarch-target-attr.cc
+++ b/src/gcc/config/loongarch/loongarch-target-attr.cc
@@ -422,51 +422,3 @@ loongarch_option_valid_attribute_p (tree fndecl, tree, tree args, int)
   return ret;
 }
 
-/* Hook to validate the current #pragma GCC target and set the state, and
-   update the macros based on what was changed.  If ARGS is NULL, then
-   POP_TARGET is used to reset the options.  */
-
-static bool
-loongarch_pragma_target_parse (tree args, tree pop_target)
-{
-  /* If args is not NULL then process it and setup the target-specific
-     information that it specifies.  */
-  if (args)
-    {
-      if (!loongarch_process_target_attr (args, NULL))
-	return false;
-
-      loongarch_option_override_internal (&la_target,
-					  &global_options,
-					  &global_options_set);
-    }
-
-  /* args is NULL, restore to the state described in pop_target.  */
-  else
-    {
-      pop_target = pop_target ? pop_target : target_option_default_node;
-      cl_target_option_restore (&global_options, &global_options_set,
-				TREE_TARGET_OPTION (pop_target));
-    }
-
-  target_option_current_node
-    = build_target_option_node (&global_options, &global_options_set);
-
-  loongarch_reset_previous_fndecl ();
-
-  /* If we're popping or reseting make sure to update the globals so that
-     the optab availability predicates get recomputed.  */
-  if (pop_target)
-    loongarch_save_restore_target_globals (pop_target);
-
-  return true;
-}
-
-/* Implement REGISTER_TARGET_PRAGMAS.  */
-
-void
-loongarch_register_pragmas (void)
-{
-  /* Update pragma hook to allow parsing #pragma GCC target.  */
-  targetm.target_option.pragma_parse = loongarch_pragma_target_parse;
-}
-- 
2.47.3

